webapp/src/marked.tsx

```diff
@@ -0,0 +1,138 @@
+interface MarkedContentState {
+}
+
+export class MarkedContent extends data.Component<MarkedContentProps, MarkedContentState> {
+
+    // Local cache for images, cleared when we create a new project.
+    // Stores code => data-uri image of decompiled result
+    private static blockSnippetCache: pxt.Map<string> = {};
+    public static clearBlockSnippetCache() {
+        this.blockSnippetCache = {};
+    }
+
+    private getBuiltinMacros() {
+        const params: pxt.Map<string> = {};
+        const theme = pxt.appTarget.appTheme;
+        if (theme.boardName)
+            params["boardname"] = pxt.Util.htmlEscape(theme.boardName);
+        if (theme.boardNickname)
+            params["boardnickname"] = pxt.Util.htmlEscape(theme.boardNickname);
+        if (theme.driveDisplayName)
+            params["drivename"] = pxt.Util.htmlEscape(theme.driveDisplayName);
+        if (theme.homeUrl)
+            params["homeurl"] = pxt.Util.htmlEscape(theme.homeUrl);
+        params["targetid"] = theme.id || "???";
+        params["targetname"] = theme.name || "Microsoft MakeCode";
+        params["targetlogo"] = theme.docsLogo ? `<img aria-hidden="true" role="presentation" class="ui mini image" src="${theme.docsLogo}" />` : "";
+        return params;
+    }
+
+    private renderSnippets(content: HTMLElement) {
+        const { parent } = this.props;
+
+        pxt.Util.toArray(content.querySelectorAll(`.lang-blocks`))
+            .forEach((langBlock: HTMLElement) => {
+                const code = langBlock.innerText;
+                const wrapperDiv = document.createElement('div');
+                langBlock.innerHTML = '';
+                langBlock.appendChild(wrapperDiv);
+                wrapperDiv.className = 'ui segment raised loading';
+                if (MarkedContent.blockSnippetCache[code]) {
+                    // Use cache
+                    const imgDiv = document.createElement('img');
+                    imgDiv.src = MarkedContent.blockSnippetCache[code];
+                    wrapperDiv.appendChild(imgDiv);
+                    pxsim.U.removeClass(wrapperDiv, 'loading');
+                } else {
+                    parent.renderBlocksAsync({
+                        action: "renderblocks", ts: code
+                    } as pxt.editor.EditorMessageRenderBlocksRequest)
+                        .done((datauri) => {
+                            if (datauri) {
+                                MarkedContent.blockSnippetCache[code] = datauri;
+                                const imgDiv = document.createElement('img');
+                                imgDiv.src = MarkedContent.blockSnippetCache[code];
+                                wrapperDiv.appendChild(imgDiv);
+                                pxsim.U.removeClass(wrapperDiv, 'loading');
+                            } else {
+                                // An error occured, show alternate message
+                                const textDiv = document.createElement('span');
```

Comment:
Show original content?

Reply:
I guess, just show Javascript.

Reply:
that way we know there's an issue, but the user can carry on perhaps?

Reply:
unlikely.